patchwork/context.py

```diff
@@ -113,6 +113,21 @@ def is_own_ancestor(self, db: Datastore) -> bool:
             context = context.parent
         return False
 
+    def can_fulfill_promise(self, db: Datastore, promise: Address) -> bool:
+        workspace = db.dereference(self.workspace_link)
+        return promise in [workspace.answer_promise,
+                           workspace.final_workspace_promise]
+
+    def can_advance_promise(self, db: Datastore, promise: Address) -> bool:
+        workspace = db.dereference(self.workspace_link)
+        promisee_contexts = \
+            (Context(args[0], db)
+             for args in db.get_promisees(workspace.answer_promise)
+                    + db.get_promisees(workspace.final_workspace_promise))
+
+        return self.can_fulfill_promise(db, promise) or \
+               any(pc.can_advance_promise(db, promise)
+                   for pc in promisee_contexts)
```

Comment:
Logic looks good. I wonder if we should factor out `workspace.answer_promise` and `workspace.final_workspace_promise` into some common construct. Not sure what it would be called though (`final_promises`? `return_promises`? `result_promises`? `output_promises`?).

Reply:
Actually it shouldn't be necessary to have promisees for both, since they get fulfilled within a single action. I have little time now, so I will look into that tomorrow.

Reply:
I would have to change too much code to unite the promises. 4ee6b9a7 follows your suggestion instead.